File: app/src/main/java/com/vocaby/app/database/DataManager.java
package com.vocaby.app.database;

import android.content.Context;
import android.util.Log;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.LinkedList;
import java.util.List;

public class DataManager {
    private static DataManager dataManager = null;
    private static Context ctx;
    private static final String HISTORY_DATA_FILE_NAME = "hVocaby";
    private static List<String> history;

    private DataManager(Context context) {
        ctx = context.getApplicationContext();
        File historyFile = new File(ctx.getFilesDir(), HISTORY_DATA_FILE_NAME);
        if(historyFile.exists()) {
            try(FileInputStream fis = ctx.openFileInput(HISTORY_DATA_FILE_NAME)) {
                ObjectInputStream ois = new ObjectInputStream(fis);
                history = (List<String>) ois.readObject();
            } catch (IOException | ClassNotFoundException e) {
                Log.d("DataManager", "Something went wrong in getInstance");
            }
        } else {
            history = new LinkedList<>();
            try (FileOutputStream fos = ctx.openFileOutput(HISTORY_DATA_FILE_NAME, Context.MODE_PRIVATE)) {
                ObjectOutputStream oos = new ObjectOutputStream(fos);
                oos.writeObject(history);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }


    public static DataManager getInstance(Context context) {
        if(dataManager == null) {
            dataManager = new DataManager(context);
        }

        return dataManager;
    }

    public List<String> writeHistory(String word) {
        history.add(0, word);
        if(history.size() > 3) {
            history.remove(3);
        }

        new Thread(() -> {
            try (FileOutputStream fos = ctx.openFileOutput(HISTORY_DATA_FILE_NAME, Context.MODE_PRIVATE)) {
                ObjectOutputStream oos = new ObjectOutputStream(fos);
                oos.writeObject(history);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }).start();

        return history;
    }

    public List<String> getHistory() {
        return history;
    }
}

File: app/src/main/java/com/vocaby/application/feature_dictionary/data/DataManager.kt
package com.vocaby.application.feature_dictionary.data

import android.content.Context
import android.util.Log
import com.vocaby.application.feature_dictionary.domain.model.SimpleEntryModel
import java.io.File
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.util.*

class DataManager(context: Context) {
    private val historyDataFileName = "hVocaby"
    private val ctx: Context = context.applicationContext
    var history: LinkedList<SimpleEntryModel>? = null
    var numHistoryItems = 6

    init {
        val historyFile = File(ctx.filesDir, historyDataFileName)
        if (historyFile.exists()) {
            try {
                ctx.openFileInput(historyDataFileName).use { fis ->
                    val ois = ObjectInputStream(fis)
                    history = ois.readObject() as LinkedList<SimpleEntryModel>?
                    ois.close()
                }
            } catch (e: IOException) {
                Log.d("DataManager", "Something went wrong in getInstance")
            } catch (e: ClassNotFoundException) {
                Log.d("DataManager", "Something went wrong in getInstance")
            }
        } else {
            history = LinkedList()
            history.let { newList ->
                try {
                    ctx.openFileOutput(historyDataFileName, Context.MODE_PRIVATE).use { fos ->
                        val oos = ObjectOutputStream(fos)
                        oos.writeObject(newList)
                        oos.close()
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }

        }
    }

    fun writeHistory(entry: SimpleEntryModel): List<SimpleEntryModel>? {
        history?.let { list ->
            list.add(0, entry)

            if (list.size > numHistoryItems) {
                list.removeAt(numHistoryItems)
            }

            try {
                ctx.openFileOutput(historyDataFileName, Context.MODE_PRIVATE).use { fos ->
                    val oos = ObjectOutputStream(fos)
                    oos.writeObject(list)
                    oos.close()
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }

        return history
    }

    fun clearHistory(): List<SimpleEntryModel>? {
        history = LinkedList()
        history?.let { newList ->
            try {
                ctx.openFileOutput(historyDataFileName, Context.MODE_PRIVATE).use { fos ->
                    val oos = ObjectOutputStream(fos)
                    oos.writeObject(newList)
                    oos.close()
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }

        return history
    }

    fun getHistory(position: Int): String? {
        return history?.get(position)?.entry
    }
}